### DISCUSSION TITLE
ASCII的歷史與應用

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS4719-C/Python-Community-GS4719-C#180] [General]
# ASCII的源起
## ASCII代表的意義
ASCII為**A**merican **S**tandard **C**ode for **I**nformation **I**nterchange的縮寫(abbreviation)
## 為何發明ASCII？
被設計使用於以7-bit方式運作(7-bit based)的「電傳打字機(teleprinter)」
![Fernscheiber_01](https://user-images.githubusercontent.com/86505454/144544616-d4cd671c-3d4c-4252-bdd2-2990fcf8cc5b.jpg)
### 何謂以7-bit方式運作
每一個ASCII碼，都能以7個bit表示。

**Comment by enoch-lightning** [comment]
#### 名詞解釋
binary的中文叫「二進制。」「二進制」是每一位加到2就要往前進一位，也就是前面那一位加1的計算方式。
bit的中文叫「位元」，是儲存二進制數字的硬體單位。
byte的中文叫「位元組。」1byte=8bit或寫成1byte=8bits(bits就是bit的複數型式。「複數」代表bit有兩個以上：2個bit、3個bit、4個bit、5個bit、6個bit或更多個bit)。1byte是電腦儲存單位中，可以被解讀出意義的最小單位。

**Reply by xCk27x** [reply]
在python中 布林值(bool)是int的子型態
但在某些語言是獨立的資料型態
請問布林值所占用的記憶體空間也是1byte嗎?
因為只有0跟1兩種數值 感覺用到1byte很佔空間

**Reply by enoch-lightning** [reply]
我目前還沒有答案，我必須先查一些資料。

**Reply by ThomasMeng0515** [reply]
> 在python中 布林值(bool)是int的子型態 但在某些語言是獨立的資料型態 請問布林值所占用的記憶體空間也是1byte嗎? 因為只有0跟1兩種數值 感覺用到1byte很佔空間

在python中布林值所佔的記憶體空間與int整數型態記憶體大小一樣

**Comment by enoch-lightning** [comment]
# 現今如何使用ASCII碼？
電腦儲存空間，如CPU、記憶體、磁碟 0 (C:)、磁碟 1 (D:)、乙太網路、GPU 0與GPU 1的最小單位是bit，但1個byte的儲存量才是儲存空間解讀意義的基本單位。所以會在二進制ASCII的七位數字前，依照下面所述規則來補成1個byte。
若二進制ASCII有偶數個1。如0, 2, ,4或6個1，則在二進制ASCII前加上1(奇數)作為parity bit成為1byte；若二進制ASCII有奇數個1。如1, 3, 5或7個1，則在二進制ASCII前加上0(偶數)作為parity bit成為1byte。
![image](https://user-images.githubusercontent.com/86505454/144554781-8541ab8a-631e-4864-97e1-eaa18b760b48.png)

**Comment by enoch-lightning** [comment]
#### 名詞解釋
偶數：「偶數」就是能被2整除的整數，也就是除以2餘數為0的整數
![image](https://user-images.githubusercontent.com/86505454/144553140-5857261b-da62-4a44-98db-57b841373dc2.png)
![image](https://user-images.githubusercontent.com/86505454/144553157-77b347b1-8fe4-491a-ad28-32e261560d7d.png)

執行程式順序，由左至右代表由先到後：1, 2, 3, 4, 5, 6, 7

**Reply by enoch-lightning** [reply]
「負數」的偶數如下
![image](https://user-images.githubusercontent.com/86505454/144553962-86c169df-749a-4260-850d-c2776cf1f983.png)
![image](https://user-images.githubusercontent.com/86505454/144553972-7608f1ba-81a5-439c-899f-d2cbc2fe0345.png)
執行程式順序，由左至右代表由先到後：1, 2, 3, 4, 5, 6

**Comment by enoch-lightning** [comment]
#### 名詞解釋
奇數：不能被2整除的整數，也就是除以2餘數為1的整數。
![image](https://user-images.githubusercontent.com/86505454/144552846-1011a26b-7ad5-4ca2-b666-26b2c8e9f01b.png)
程式執行順序，由左至右代表由先到後：1, 2, 3, 4

**Reply by enoch-lightning** [reply]
「負數」的「奇數」如下：
![image](https://user-images.githubusercontent.com/86505454/144554476-9c72e31c-4d53-444d-943a-dbc8b284f4b3.png)
![image](https://user-images.githubusercontent.com/86505454/144554487-6cf2d9c6-ea65-46c7-9e2d-4ff1a352aa7f.png)

疑惑：加上負號(-)好像對取餘數的運算，完全沒有影響？不知是怎麼回事。

**Reply by xCk27x** [reply]
事實上是會有影響的 
例如 1 % -3 會得到 -2 
原因是python除法運算遵守一個規則
**被除數 = 除數\*(被除數 // 除數) + 餘數**
因為 1 // -3 等於 -1 (整數除法 : 向下找整數)
所以根據除法規則 餘數會等於-2

**Comment by HSU-YU-JEN** [comment]
謝謝同學分享，寫各種程式時常常會用到ASCII表來判斷，但現在才真正理解他的背後原理與儲存空間有關

**Comment by pekoranoinu** [comment]
原來ASCII的原理是這樣啊!

**Comment by RogelioKG** [comment]
不知道還有沒有人看到這篇，
但小弟想以更廣義的角度，
來探討異號模除法的問題。

**被除數 ÷ 除數 = 商數 ... 餘數**
```
 a ÷ b = a//b ... a%b
```
不僅是 Python 遵守這樣的除法運算規則，
而是全世界都會遵守這樣的規則。
只不過，大家對於商數要怎麼取，似乎有不同的觀點。

相信有寫過 C/C++ 的朋友肯定有困惑過，
為何 C/C++ 的異號模除法和 Python 的結果不一樣？
### 同號除法結果相同
```
      a ÷  b = a//b ... a%b
     13 ÷  6 =    2 ...   1 (C/C++)
     13 ÷  6 =    2 ...   1 (Python)

      a ÷  b = a//b ... a%b
    -13 ÷ -6 =    2 ...  -1 (C/C++)
    -13 ÷ -6 =    2 ...  -1 (Python)
```
### 異號除法結果不同
```
      a ÷  b = a//b ... a%b
    -13 ÷  6 =   -2 ...  -1 (C/C++)
    -13 ÷  6 =   -3 ...   5 (Python)

      a ÷  b = a//b ... a%b
     13 ÷ -6 =   -2 ...   1 (C/C++)
     13 ÷ -6 =   -3 ...  -5 (Python)
 ```

(紅色代表商數，綠色代表餘數) 

不難發現，C++的商數 (整數除法) 採用了 **向 0 取整** 的方法：

![mod in C/C++](https://github.com/NCU-GS4719-Python/Python-Community/assets/86526364/4269ff31-f031-48c9-83cd-c6c0d2ead24a)

而 Python 的商數 (整數除法) 採用了 **向下取整** (floor) 的方法：

![mod in Python](https://github.com/NCU-GS4719-Python/Python-Community/assets/86526364/8c100b28-198e-44aa-b91c-515c5cf3d034)

現今數學界較常使用的方法，與 Python 相同，是採用向下取整的方法。

你可以用以下這種方式記下來。
我個人幫助我自己回想的理解是：
如果除數是 6，那麼餘數就要落在 0\~6 之間。
所以如果除數是 -6，那麼餘數就要落在 0\~-6 之間。
Python 和一般數學模除法都是這麼做的。

如果要讓 C/C++ 可以用上數學版的模除法，
你需要寫一個巨集，下面這個是我常用的巨集。
```C++
#define MOD(a,b)  (((a)%(b) + (b)) % (b)) // 修正成和 Python 一樣的模除法
```
